src/waveform.js

```diff
@@ -0,0 +1,189 @@
+              i,
+              datapoints,
+              this.waveformColour
+            );
+          }
+        }
+      );
+    }
+  }
+
+  static toggleWaveformCanvasCallback(event) {
+    this.log.info("waveform toggle: " + event.target.checked);
+    this.canvas.style.display = event.target.checked ? "" : "none";
+  }
+
+  static monitorAudioCanPlayCallback() {
+    let audio = document.querySelector("audio");
+    if (!audio.paused && this.canvasDisplayToggle.checked)
+      this.generateWaveform();
+  }
+
+  static monitorAudioTimeupdateCallback(e) {
+    let audio = e.target;
+    let progress = audio.currentTime / audio.duration;
+    Waveform.drawOverlay(
+      this.canvas,
+      progress,
+      this.waveformOverlayColour,
+      this.waveformColour
+    );
+  }
+
+  static createCanvas() {
+    let canvas = document.createElement("canvas");
+    canvas.style.display = "none";
+    canvas.classList.add("waveform");
+
+    // configure element to properly hold canvas
+    let progbar = document.querySelector("div.progbar");
+    progbar.classList.add("waveform");
+
+    let div = document.createElement("div");
+    div.append(canvas);
+    progbar.prepend(div);
+    return canvas;
+  }
+
+  static createCanvasDisplayToggle() {
+    let toggle = document.createElement("input");
+
+    toggle.setAttribute("title", "toggle waveform display");
+    toggle.setAttribute("type", "checkbox");
+    toggle.setAttribute("class", "waveform");
+    toggle.setAttribute("id", "switch");
+
+    let label = document.createElement("label");
+    label.setAttribute("class", "waveform");
+    label.htmlFor = "switch";
+    label.innerHTML = "Toggle";
+
+    let toggle_div = document.createElement("div");
+    toggle_div.append(toggle);
+    toggle_div.append(label);
+
+    let inlineplayer = document.querySelector("div.controls");
+    inlineplayer.append(toggle_div);
+
+    return toggle;
+  }
+
+  static fillBar(canvas, amplitude, index, numElements, colour = "white") {
+    let ctx = canvas.getContext("2d");
+    ctx.globalCompositeOperation = "source-over";
+    ctx.fillStyle = colour;
+
+    let graphHeight = canvas.height * amplitude;
+    let barWidth = canvas.width / numElements;
+    let position = index * barWidth;
+    ctx.fillRect(position, canvas.height, barWidth, -graphHeight);
+  }
+
+  static drawOverlay(canvas, progress, colour = "red", clearColour = "black") {
+    let ctx = canvas.getContext("2d");
+    ctx.globalCompositeOperation = "source-atop";
+    ctx.fillStyle = clearColour;
+    ctx.fillRect(0, 0, canvas.width, canvas.height);
+    ctx.fillStyle = colour;
+    ctx.fillRect(0, 0, canvas.width * progress, canvas.height);
+  }
+
+  static invertColour(colour) {
+    let rgb = colour
+      .split("rgb(")[1]
+      .split(")")[0]
+      .split(",");
+
+    let r = parseInt(255 - rgb[0]);
+    let g = parseInt(255 - rgb[1]);
+    let b = parseInt(255 - rgb[2]);
+
+    return "rgb(" + r + "," + g + "," + b + ")";
```

Comment:
Paraphrased for ES6:

```js
return `rgb(${r}, ${g}, ${b})`
```

Kinda surprised there isn't a nicer colour enum or struct by now. Carry on 👍